File: week4/python/LRUCache_146.py
class Node:
    def __init__(self, key: int, value: int):
        self.key = key
        self.val = value
        self.next = None
        self.prev = None
        
        
class LRUCache:
    
    def __init__(self, capacity: int):
        self.head = Node(-1,-1)
        self.tail = Node(-1,-1)
        self.head.next = self.tail
        self.tail.prev = self.head
        self.record = dict()
        self.capacity = capacity
        
    def get(self, key: int) -> int:
        if key in self.record:
            node = self.record.get(key)
            self.updateRecord(node)
            return node.val
        else:
            return -1
        

    def put(self, key: int, value: int) -> None:
        if self.capacity == 0:
            return
        
        if key in self.record:
            node = self.record.get(key)
            node.val = value
            self.updateRecord(node)
        else:
            if self.capacity == len(self.record):
                node = self.head.next
                self.removeRecord(node)
                self.record.pop(node.key)
                
            node = Node(key,value)
            self.putRecord(node)
            self.record[key] = node
        
    def removeRecord(self, node: Node) -> None:
        left, right = node.prev, node.next
        left.next = right
        right.prev = left
        
        
    def putRecord(self, node: Node) -> None:
        left, right = self.tail.prev, self.tail
        left.next = node
        node.prev = left
        right.prev = node
        node.next = right
        
        
    def updateRecord(self, node: Node) -> None:
        self.removeRecord(node)
        self.putRecord(node)
        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: week4/java/LRUCache_146.java
class LRUCache {

    class Node {
        int key;
        int val;
        Node next, prev;
        Node(int _key, int _val) {
            key = _key;
            val = _val;
        }
    }
    
    private Node head, tail;
    private int maxCapacity;
    private Map<Integer, Node> record;
    
    public LRUCache(int capacity) {
        head = new Node(-1, -1);
        tail = new Node(-1, -1);
        head.next = tail;
        tail.prev = head;
        maxCapacity = capacity;
        record = new HashMap<>();
    }
    
    public int get(int key) {
        int value = -1;
        if (record.containsKey(key)) {
            Node node = record.get(key);
            updateRecord(node);
            value = node.val;
        }
        return value;
    }
    
    public void put(int key, int value) {
        if (record.containsKey(key)) {
            Node node = record.get(key);
            node.val = value;
            updateRecord(node);
        } else {
            if (maxCapacity == 0) {
                return;
                
            } else if (maxCapacity <= record.size()) {
                record.remove(tail.prev.key);
                removeRecord(tail.prev);
            }
            
            Node newNode = new Node(key, value);
            putRecord(newNode);
        }
    }
    
    private Node removeRecord(Node removedNode) {
        removedNode.prev.next = removedNode.next;
        removedNode.next.prev = removedNode.prev;
        removedNode.next = null;
        removedNode.prev = null;
        return removedNode;
    }
    
    private void putRecord(Node node) {
        head.next.prev = node;
        node.next = head.next;
        node.prev = head;
        head.next = node;
        record.put(node.key, node);
    }
    
    private void updateRecord(Node node) {
        removeRecord(node);
        putRecord(node);
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */
